File: backend/app/schemas/comparisons.py
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict


class ScanComparisonCreate(BaseModel):
    baseline_scan_id: int
    target_scan_id: int


class ScanComparisonBuildRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    scan_comparison_id: int
    comparison_version: str
    algorithm_identity: str
    status: str
    baseline_projection_build_id: int | None
    target_projection_build_id: int | None
    baseline_projection_version: str | None
    target_projection_version: str | None
    baseline_projection_algorithm_identity: str | None
    target_projection_algorithm_identity: str | None
    baseline_projection_checksum: str | None
    target_projection_checksum: str | None
    baseline_scope_fingerprint: str | None
    target_scope_fingerprint: str | None
    baseline_seed_fingerprint: str | None
    target_seed_fingerprint: str | None
    coverage_state: str | None
    warnings_json: list[str]
    validation_json: dict[str, Any]
    comparison_checksum_sha256: str | None
    started_at: datetime | None
    finished_at: datetime | None
    failed_at: datetime | None
    build_duration_ms: int | None
    error_type: str | None
    error_message: str | None
    page_result_count: int
    resource_result_count: int
    link_result_count: int
    created_at: datetime


class ComparisonScanRead(BaseModel):
    id: int
    status: str
    starting_url: str
    created_at: datetime
    started_at: datetime | None
    finished_at: datetime | None
    stop_reason: str | None
    failed_count: int


class ScanComparisonRead(BaseModel):
    id: int
    website_property_id: int
    baseline_scan_id: int
    target_scan_id: int
    current_build_id: int | None
    created_at: datetime
    updated_at: datetime
    baseline_scan: ComparisonScanRead
    target_scan: ComparisonScanRead
    current_build: ScanComparisonBuildRead | None
    active_build: ScanComparisonBuildRead | None = None


class ScanComparisonList(BaseModel):
    items: list[ScanComparisonRead]
    total: int
    limit: int
    offset: int


class ComparisonPageRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    resource_id: int
    normalized_url: str
    host: str
    path: str
    baseline_page_projection_id: int | None
    target_page_projection_id: int | None
    baseline_snapshot_id: int | None
    target_snapshot_id: int | None
    presence_state: str
    baseline_presence_detail: str
    target_presence_detail: str
    change_state: str
    content_state: str
    head_state: str
    changed_field_count: int
    content_changed: bool
    head_changed: bool
    http_status_changed: bool
    fetch_state_changed: bool
    final_url_changed: bool
    redirect_state_changed: bool
    content_type_changed: bool
    title_changed: bool
    canonical_changed: bool
    robots_changed: bool
    language_changed: bool
    depth_changed: bool
    inbound_links_changed: bool
    outbound_links_changed: bool
    embedded_resources_changed: bool
    rendered_state_changed: bool
    rendered_counts_changed: bool
    baseline_http_status: int | None
    target_http_status: int | None
    baseline_content_hash: str | None
    target_content_hash: str | None
    baseline_head_hash: str | None
    target_head_hash: str | None
    response_time_ms_delta: int | None
    network_bytes_delta: int | None
    raw_html_size_delta: int | None
    stored_html_size_delta: int | None
    outgoing_edges_newly_observed: int
    outgoing_edges_not_observed: int
    outgoing_edges_changed: int
    incoming_edges_newly_observed: int
    incoming_edges_not_observed: int
    incoming_edges_changed: int
    baseline_json: dict[str, Any] | None
    target_json: dict[str, Any] | None


class ComparisonPageList(BaseModel):
    items: list[ComparisonPageRead]
    total: int
    limit: int
    offset: int
    comparison_build_id: int
    comparison_version: str


class ComparisonResourceRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    resource_id: int
    normalized_url: str
    host: str
    path: str
    baseline_snapshot_id: int | None
    target_snapshot_id: int | None
    presence_state: str
    change_state: str
    changed_field_count: int
    baseline_kind: str | None
    target_kind: str | None
    baseline_mime_type: str | None
    target_mime_type: str | None
    baseline_http_status: int | None
    target_http_status: int | None
    status_changed: bool
    observed_state_changed: bool
    occurrence_delta: int | None
    source_page_delta: int | None
    declared_size_delta: int | None
    baseline_json: dict[str, Any] | None
    target_json: dict[str, Any] | None


class ComparisonResourceList(BaseModel):
    items: list[ComparisonResourceRead]
    total: int
    limit: int
    offset: int
    comparison_build_id: int
    comparison_version: str


class ComparisonLinkRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    source_resource_id: int
    target_resource_id: int
    source_url: str
    target_url: str
    baseline_source_snapshot_id: int | None
    target_source_snapshot_id: int | None
    presence_state: str
    change_state: str
    changed_field_count: int
    baseline_occurrence_count: int
    target_occurrence_count: int
    occurrence_delta: int
    baseline_json: dict[str, Any] | None
    target_json: dict[str, Any] | None


class ComparisonLinkList(BaseModel):
    items: list[ComparisonLinkRead]
    total: int
    limit: int
    offset: int
    comparison_build_id: int
    comparison_version: str


class ScanComparisonOverview(BaseModel):
    comparison: ScanComparisonRead
    summary: dict[str, Any] | None


class OccurrenceDiffRead(BaseModel):
    state: Literal["present_in_both", "newly_observed", "not_observed_in_target"]
    fingerprint: str
    occurrence: dict[str, Any]
    count: int


class OccurrenceDiffList(BaseModel):
    items: list[OccurrenceDiffRead]
    total: int
    limit: int
    offset: int
    compared_baseline_count: int
    compared_target_count: int
    truncated: bool


class SourceDiffRead(BaseModel):
    state: Literal[
        "available",
        "identical",
        "baseline_missing",
        "target_missing",
        "too_large",
        "decoding_failed",
        "truncated",
    ]
    diff_text: str
    input_truncated: bool = False
    output_truncated: bool = False


class PageChangeHistoryItem(BaseModel):
    scan_id: int
    snapshot_id: int
    scan_created_at: datetime
    scan_status: str
    observed_at: datetime | None
    http_status: int | None
    fetch_state: str
    content_hash: str | None
    head_hash: str | None
    title: str | None
    canonical_url: str | None
    robots_directives: str | None
    rendered_state: str | None
    change_label: str
    changed_flags: list[str]
    previous_snapshot_id: int | None
    previous_scan_id: int | None
    intervening_scan_count: int
    intervening_unsuccessful_observation_count: int


class PageChangeHistoryList(BaseModel):
    items: list[PageChangeHistoryItem]
    total: int
    limit: int
    offset: int

File: frontend/src/types/comparisons.ts
export type ComparisonScan = { id: number; status: string; starting_url: string; created_at: string; started_at: string | null; finished_at: string | null; stop_reason: string | null; failed_count: number };

export type ScanComparisonBuild = {
  id: number; scan_comparison_id: number; comparison_version: string; algorithm_identity: string; status: string;
  baseline_projection_build_id: number | null; target_projection_build_id: number | null;
  baseline_projection_version: string | null; target_projection_version: string | null;
  baseline_projection_algorithm_identity: string | null; target_projection_algorithm_identity: string | null;
  baseline_projection_checksum: string | null; target_projection_checksum: string | null;
  baseline_scope_fingerprint: string | null; target_scope_fingerprint: string | null;
  baseline_seed_fingerprint: string | null; target_seed_fingerprint: string | null;
  coverage_state: string | null; warnings_json: string[]; validation_json: Record<string, unknown>;
  comparison_checksum_sha256: string | null; started_at: string | null; finished_at: string | null;
  failed_at: string | null; build_duration_ms: number | null; error_type: string | null; error_message: string | null;
  page_result_count: number; resource_result_count: number; link_result_count: number; created_at: string;
};

export type ScanComparison = {
  id: number; website_property_id: number; baseline_scan_id: number; target_scan_id: number;
  current_build_id: number | null; created_at: string; updated_at: string;
  baseline_scan: ComparisonScan; target_scan: ComparisonScan;
  current_build: ScanComparisonBuild | null; active_build: ScanComparisonBuild | null;
};

export type ScanComparisonJobProgress = {
  id: number; status: string; current_operation: string | null;
  progress_current: number | null; progress_total: number | null; progress_unit: string | null;
  started_at: string | null; heartbeat_at: string | null;
};

export type ScanComparisonOverview = { comparison: ScanComparison; summary: { pages: Record<string, number>; resources: Record<string, number>; links: Record<string, number>; scan: Record<string, unknown> } | null; active_job?: ScanComparisonJobProgress | null };
export type ScanComparisonList = { items: ScanComparison[]; total: number; limit: number; offset: number };

export type ComparisonPage = {
  id: number; resource_id: number; normalized_url: string; host: string; path: string;
  baseline_page_projection_id: number | null; target_page_projection_id: number | null;
  baseline_snapshot_id: number | null; target_snapshot_id: number | null;
  presence_state: string; baseline_presence_detail: string; target_presence_detail: string;
  change_state: string; content_state: string; head_state: string; changed_field_count: number;
  exact_source_state: string; exact_source_changed: boolean;
  baseline_normalized_source_hash: string | null; target_normalized_source_hash: string | null;
  normalized_source_state: string; document_content_state: string; metadata_state: string; technical_state: string;
  primary_change_class: string; normalization_only_changed: boolean;
  source_difference_categories_json: string[]; normalization_details_json: Array<Record<string, unknown>>;
  content_changed: boolean; head_changed: boolean; http_status_changed: boolean; fetch_state_changed: boolean;
  final_url_changed: boolean; redirect_state_changed: boolean; content_type_changed: boolean; title_changed: boolean;
  canonical_changed: boolean; robots_changed: boolean; language_changed: boolean; depth_changed: boolean;
  inbound_links_changed: boolean; outbound_links_changed: boolean; embedded_resources_changed: boolean;
  baseline_http_status: number | null; target_http_status: number | null;
  baseline_content_hash: string | null; target_content_hash: string | null;
  baseline_head_hash: string | null; target_head_hash: string | null;
  response_time_ms_delta: number | null; network_bytes_delta: number | null;
  raw_html_size_delta: number | null; stored_html_size_delta: number | null;
  outgoing_edges_newly_observed: number; outgoing_edges_not_observed: number; outgoing_edges_changed: number;
  incoming_edges_newly_observed: number; incoming_edges_not_observed: number; incoming_edges_changed: number;
  baseline_json: Record<string, unknown> | null; target_json: Record<string, unknown> | null;
};

export type ComparisonResource = {
  id: number; resource_id: number; normalized_url: string; host: string; path: string;
  baseline_snapshot_id: number | null; target_snapshot_id: number | null; presence_state: string; change_state: string;
  changed_field_count: number; baseline_kind: string | null; target_kind: string | null;
  baseline_mime_type: string | null; target_mime_type: string | null;
  baseline_http_status: number | null; target_http_status: number | null; status_changed: boolean;
  observed_state_changed: boolean; occurrence_delta: number | null; source_page_delta: number | null;
  declared_size_delta: number | null; baseline_json: Record<string, unknown> | null; target_json: Record<string, unknown> | null;
};

export type ComparisonLink = {
  id: number; source_resource_id: number; target_resource_id: number; source_url: string; target_url: string;
  baseline_source_snapshot_id: number | null; target_source_snapshot_id: number | null;
  presence_state: string; change_state: string; changed_field_count: number;
  baseline_occurrence_count: number; target_occurrence_count: number; occurrence_delta: number;
  baseline_json: Record<string, unknown> | null; target_json: Record<string, unknown> | null;
};

export type ComparisonResultList<T> = { items: T[]; total: number; limit: number; offset: number; comparison_build_id: number; comparison_version: string };
export type SourceDiff = { state: string; diff_text: string; mode: "exact" | "meaningful"; input_truncated: boolean; output_truncated: boolean };
export type OccurrenceDiff = { items: Array<{ state: string; fingerprint: string; occurrence: Record<string, unknown>; count: number }>; total: number; limit: number; offset: number; compared_baseline_count: number; compared_target_count: number; truncated: boolean };
export type PageChangeHistory = { items: Array<{ scan_id: number; snapshot_id: number; scan_created_at: string; scan_status: string; observed_at: string | null; http_status: number | null; fetch_state: string; content_hash: string | null; head_hash: string | null; title: string | null; canonical_url: string | null; robots_directives: string | null; change_label: string; changed_flags: string[]; previous_snapshot_id: number | null; previous_scan_id: number | null; intervening_scan_count: number; intervening_unsuccessful_observation_count: number }>; total: number; limit: number; offset: number };
